core/src/main/java/io/questdb/griffin/engine/join/AsOfJoinMemoizedRecordCursorFactory.java

```diff
@@ -0,0 +1,355 @@
+    private final AsofJoinColumnAccessHelper columnAccessHelper;
+    private final AsOfJoinMemoizedRecordCursor cursor;
+    private final int slaveSymbolColumnIndex;
+    private final long toleranceInterval;
+
+    public AsOfJoinMemoizedRecordCursorFactory(
+            CairoConfiguration configuration,
+            RecordMetadata metadata,
+            RecordCursorFactory masterFactory,
+            RecordCursorFactory slaveFactory,
+            int columnSplit,
+            int slaveSymbolColumnIndex,
+            AsofJoinColumnAccessHelper columnAccessHelper,
+            JoinContext joinContext,
+            long toleranceInterval
+    ) {
+        super(metadata, joinContext, masterFactory, slaveFactory);
+        assert slaveFactory.supportsTimeFrameCursor();
+        this.columnAccessHelper = columnAccessHelper;
+        this.toleranceInterval = toleranceInterval;
+        this.slaveSymbolColumnIndex = slaveSymbolColumnIndex;
+        this.cursor = new AsOfJoinMemoizedRecordCursor(
+                columnSplit,
+                NullRecordFactory.getInstance(slaveFactory.getMetadata()),
+                masterFactory.getMetadata().getTimestampIndex(),
+                masterFactory.getMetadata().getTimestampType(),
+                slaveFactory.getMetadata().getTimestampIndex(),
+                slaveFactory.getMetadata().getTimestampType(),
+                configuration.getSqlAsOfJoinLookAhead()
+        );
+    }
+
+    @Override
+    public boolean followedOrderByAdvice() {
+        return masterFactory.followedOrderByAdvice();
+    }
+
+    @Override
+    public RecordCursor getCursor(SqlExecutionContext executionContext) throws SqlException {
+        RecordCursor masterCursor = masterFactory.getCursor(executionContext);
+        TimeFrameRecordCursor slaveCursor = null;
+        try {
+            slaveCursor = slaveFactory.getTimeFrameCursor(executionContext);
+            cursor.of(masterCursor, slaveCursor, executionContext.getCircuitBreaker());
+            return cursor;
+        } catch (Throwable e) {
+            Misc.free(slaveCursor);
+            Misc.free(masterCursor);
+            throw e;
+        }
+    }
+
+    @Override
+    public int getScanDirection() {
+        return masterFactory.getScanDirection();
+    }
+
+    @Override
+    public boolean recordCursorSupportsRandomAccess() {
+        return false;
+    }
+
+    @Override
+    public void toPlan(PlanSink sink) {
+        sink.type("AsOf Join Memoized Scan");
+        sink.attr("condition").val(joinContext);
+        sink.child(masterFactory);
+        sink.child(slaveFactory);
+    }
+
+    @Override
+    protected void _close() {
+        Misc.freeIfCloseable(getMetadata());
+        Misc.free(masterFactory);
+        Misc.free(slaveFactory);
+    }
+
+    private class AsOfJoinMemoizedRecordCursor extends AbstractKeyedAsOfJoinRecordCursor {
+
+        private static final long NOT_REMEMBERED = Long.MIN_VALUE;
```

Comment:
Usage of this constant looks weird: it's written in a single place and never used in any checks. Is this on purpose?

Reply:
It didn't need checking because another check ensured the stored map value was NOT_REMEMBERED, and other places exploit it being the min value of long. After some more changes, we now have an explicit check as well.